File: app/src/main/java/com/github/uryyyyyyy/androidsample/MainActivity.java
package com.github.uryyyyyyy.androidsample;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.GoogleAuthProvider;

import static android.provider.AlarmClock.EXTRA_MESSAGE;

public class MainActivity extends AppCompatActivity {

    private FirebaseAuth mAuth;
    private FirebaseAuth.AuthStateListener mAuthListener;
    private static final String TAG = "MainActivity";
    public static final int ACTIVITY_FLAG = 2;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        LinearLayout layout = new LinearLayout(this);

        Button jumpButton = new Button(this);
        jumpButton.setText("Jump");
        jumpButton.setLayoutParams(new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT)
        );
        jumpButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                jumpToActivity(v);
            }
        });

        layout.addView(jumpButton);
        setContentView(layout);

        mAuth = FirebaseAuth.getInstance();

        mAuthListener = new FirebaseAuth.AuthStateListener() {
            @Override
            public void onAuthStateChanged(@NonNull FirebaseAuth firebaseAuth) {
                FirebaseUser user = firebaseAuth.getCurrentUser();
                if (user != null) {
                    // User is signed in
                    Log.d(TAG, "onAuthStateChanged:signed_in:" + user.getUid());
                } else {
                    // User is signed out
                    Log.d(TAG, "onAuthStateChanged:signed_out");
                }
                // ...
            }
        };
    }

    private void jumpToActivity(View v){
        Intent intent = new Intent(this, DisplayMessageActivity.class);
        String message = "loginしてください";
        intent.putExtra(EXTRA_MESSAGE, message);
        startActivityForResult(intent, ACTIVITY_FLAG);
    }

    @Override
    public void onStart() {
        super.onStart();
        mAuth.addAuthStateListener(mAuthListener);
    }

    @Override
    public void onStop() {
        super.onStop();
        if (mAuthListener != null) {
            mAuth.removeAuthStateListener(mAuthListener);
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        Log.d(TAG, "onActivityResult");

        if (requestCode == ACTIVITY_FLAG) {
            String googleAccountId = data.getStringExtra("google.id");
            String googleAccountIdToken = data.getStringExtra("google.idToken");

            Log.d(TAG, "firebaseAuthWithGoogle:" + googleAccountId);

            AuthCredential credential = GoogleAuthProvider.getCredential(googleAccountIdToken, null);
            mAuth.signInWithCredential(credential)
                    .addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                        @Override
                        public void onComplete(@NonNull Task<AuthResult> task) {
                            Log.d(TAG, "signInWithCredential:onComplete:" + task.isSuccessful());

                            // If sign in fails, display a message to the user. If sign in succeeds
                            // the auth state listener will be notified and logic to handle the
                            // signed in user can be handled in the listener.
                            if (!task.isSuccessful()) {
                                Log.w(TAG, "signInWithCredential", task.getException());
                                Toast.makeText(MainActivity.this, "Authentication failed.",
                                        Toast.LENGTH_SHORT).show();
                            }
                            // ...
                        }
                    });
        }
    }

}

File: app/src/main/kotlin/com/github/uryyyyyyy/androidsample/MainActivity.kt
package com.github.uryyyyyyy.androidsample;

import android.app.Activity
import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import io.reactivex.subjects.BehaviorSubject


class MainActivity : Activity() {

    val count: BehaviorSubject<Int> = BehaviorSubject.createDefault(0)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val linearLayout = LinearLayout(this)
        linearLayout.orientation = LinearLayout.VERTICAL
        setContentView(linearLayout)

        val addButton = Button(this)
        addButton.text = "add 1"
        linearLayout.addView(addButton)
        addButton.setOnClickListener{Unit ->
            count.onNext(count.value + 1)
        }

        val minusButton = Button(this)
        minusButton.text = "minus 1"
        linearLayout.addView(minusButton)
        minusButton.setOnClickListener{Unit ->
            count.onNext(count.value - 1)
        }

        val scoreView = TextView(this)
        linearLayout.addView(scoreView)
        count.subscribe{v -> scoreView.text = "${v}点だよ" }

        val backButton = Button(this)
        backButton.text = "back"
        linearLayout.addView(backButton)
//        backButton.setOnClickListener{v ->
//            val intent = Intent(this, javsaClass<MainActivity>())
//            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP + Intent.FLAG_ACTIVITY_SINGLE_TOP
//            startActivity(intent)
//        }
        Log.e("shiba", "last")
    }
}